marimo/_islands/island_generator.py

```diff
@@ -0,0 +1,249 @@
+            <marimo-cell-code hidden>
+            {uri_encode_component(self.code) if include_code else ""}
+            </marimo-cell-code>
+        </marimo-island>
+        """
+            ).strip()
+        )
+
+
+class MarimoIslandGenerator:
+    """
+    Generates Marimo islands for embedding in other pages.
+
+    This is a great way to use another SSG framework that converts
+    Python code to HTML using marimo-islands.
+
+    Generally you will want to:
+
+    1. Find all the code snippets and add them to the generator.
+    2. Build the app.
+    3. Replace all code snippets with the rendered HTML.
+    4. Include the header in the <head> tag.
+
+    # Example
+
+    ```python
+    from marimo import MarimoIslandGenerator
+
+    generator = MarimoIslandGenerator()
+    block1 = generator.add_code("import marimo as mo")
+    block2 = generator.add_code("mo.md('Hello, islands!')")
+
+    # Build the app
+    app = await generator.build()
+
+    # Render the app
+    output = f\"\"\"
+    <html>
+        <head>
+            {generator.render_head()}
+        </head>
+        <body>
+            {block1.render(include_output=False)}
+            {block2.render()}
+        </body>
+    </html>
+    \"\"\"
+    ```
+    """
+
+    def __init__(self, app_id: str = "main"):
+        self.has_run = False
+        self._app_id = app_id
+        self._app = InternalApp(App())
+        self._stubs: List[MarimoIslandStub] = []
+
+    def add_code(
+        self,
+        code: str,
+    ) -> MarimoIslandStub:
+        """Add a code cell to the app.
+
+        *Args:*
+
+        - code (str): The code to add to the app.
+        """
+
+        cell_id = self._app.cell_manager.create_cell_id()
+        cell_impl = compile_cell(dedent(code), cell_id)
+        cell_impl.configure(CellConfig(disabled=False, hide_code=False))
+        cell = Cell(_name="__", _cell=cell_impl)
+
+        self._app.cell_manager._register_cell(
+            cell,
+            app=self._app,
+        )
+
+        stub = MarimoIslandStub(
+            cell_id=cell_id,
+            app_id=self._app_id,
+            code=code,
+        )
+        self._stubs.append(stub)
+
+        return stub
+
+    async def build(self) -> App:
+        """
+        Build the app. This should be called after adding all the code cells.
+
+        *Returns:*
+
+        - App: The built app.
+        """
+
+        if self.has_run:
+            raise ValueError("You can only call build() once")
+
+        session = await run_app_until_completion(
+            file_manager=AppFileManager.from_app(self._app),
+            cli_args={},
+        )
+        self.has_run = True
+
+        for stub in self._stubs:
+            stub._internal_app = self._app
+            stub._session = session
+
+        return cast(App, self._app)
+
+    def render_head(
+        self,
+        *,
+        version_override: str = __version__,
+        _development_url: bool = False,
```

Comment:
`: Optional[str] = None` ?

Maybe I'm weird, but I do most everything through a reverse proxy under *.tld or <port>.port.tld

Also happy to hack on this branch too - just let me know when you're done actively making changes

Reply:
im done making changes, feel free to add to a branch or commit directly, much apprecaited!